### PR TITLE
Full docs of the OPC UA public beta release

### PR DESCRIPTION
**Pull request by jaspervanderhoek**
Added Known limitation section
Started updating first set of chapters (up to chapter 3.2.2.)

### CODE REVIEW

**Comment by MarkvanMents** [comment]
Hi Jasper @jaspervanderhoek ,
The changes look great and add a lot of depth and background which was missing. Thanks.
Let me know when it is ready to be reviewed properly.

I notice that there has been an accidental `:%s/us/UA/g` - if you still have it open in an editor perhaps that would be the easiest place to undo it with a ^z rather than reverting the whole commit or reversing it by hand?

I look forward to hearing from you when this is ready to review, and let me know if there is anything I can help you with.
Mark

**Comment by jaspervanderhoek** [comment]
> I notice that there has been an accidental `:%s/us/UA/g` - if you still have it open in an editor perhaps that would be the easiest place to undo it with a ^z rather than reverting the whole commit or reversing it by hand?

Unfortunately I made that change a while ago (I had some typos where I typed OPC US instead of UA) apparently I forgot to check the case sensitive option with the find/replace.

I think I fixed all mistakes though.

**Comment by MarkvanMents** [comment]
Thanks for resolving so quickly! Looks fine now.

**Comment by MarkvanMents** [comment]
Hi Jasper, @jaspervanderhoek
Is this ready to be reviewed and merged, or is there more work you need to do on it?

**Comment by jaspervanderhoek** [comment]
Hi @MarkvanMents , yes it's ready to be merged. There are just 2 todo's left in the document, we need a new screenshot of the module in the project explorer. And I added a todo to do an in-page link to another chapter, I don't know how you write those links for our doc pages.

File: content/appstore/connectors/opc-ua.md [hunk 1]
```diff
@@ -79,118 +93,154 @@ The module/app is designed for usage with multiple servers if necessary. The sta
...
+* CertificatePassword (String) – the encrypted password for the certificate, automatically takes the CertificatePassword and encrypts the value
 
 You can see an example of how this can be set up in the [OPC UA Client example implementation](#example-implementation) section.
 
+
 ### 3.2 Actions
 
 Once you have set up the server configuration, you can perform the following actions in your microflows.
 
 #### 3.2.1 **Browse** a List of Nodes
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:112]:
```suggestion
#### 3.2.1 **Browse** a List of Nodes{#browse}
```

File: content/appstore/connectors/opc-ua.md [hunk 2]
```diff
@@ -219,52 +269,65 @@ This page allows you to create or change the details of an OPC UA server you wan
...
+**Test Messaging:**  
+* **NodeId** – The NodeId of the Node you want to use. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
+* **Value to write** – the value to write to a node when the **Write** button is clicked. Leave blank if you want to execute any of the other actions.
+* **Result** – the resulting JSON string from performing any of the four test actions. This shows the full response from the OPC UA Server, the connector might at relevant details in case of an error. 
 
-This displays the following data entry fields which can be used to control the tests.
+* **Read** executes the Read action {{% todo %}}[Add a link to the read-action chapter]{{% /todo %}}
+* **Write** executes the Write action {{% todo %}}[Add a link to the Write-action chapter]{{% /todo %}}
+* **Browse** executes the Browse action {{% todo %}}[Add a link to the Browse-action chapter]{{% /todo %}}
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:294]:
```suggestion
* **Browse** executes the [Browse action](#browse)
```

File: content/appstore/connectors/opc-ua.md [hunk 3]
```diff
@@ -79,118 +93,154 @@ The module/app is designed for usage with multiple servers if necessary. The sta
...
+* Opc UA server cfg – an object of entity type OpcUaServerCfg containing the configuration of the server to which the request is made
+* NodeId – The NodeId from where you want to browse to it's children. When requesting the 'Is Root'=true leave this value empty. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
 * Is root – is used by the tree widget in the example implementation module — if you are not using the tree widget you, the value here is not important
-* Use return value – `Yes` creates a variable containing the list of nodes, `No` does not return a variable
-* Variable name – the name assigned to the variable containing the return value
+* Return Variable – The output of this action is the JSON string with all information about the requested nodes. This string can be parsed with the Import Mapping activity.
+
 
 #### 3.2.2 **Read** the Value of a Node
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:124]:
```suggestion
#### 3.2.2 **Read** the Value of a Node{#read}
```

File: content/appstore/connectors/opc-ua.md [hunk 4]
```diff
@@ -219,52 +269,65 @@ This page allows you to create or change the details of an OPC UA server you wan
...
 
-To display the test buttons and additional data fields, check **Enable Test Mode**.
+**Test Messaging:**  
+* **NodeId** – The NodeId of the Node you want to use. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
+* **Value to write** – the value to write to a node when the **Write** button is clicked. Leave blank if you want to execute any of the other actions.
+* **Result** – the resulting JSON string from performing any of the four test actions. This shows the full response from the OPC UA Server, the connector might at relevant details in case of an error. 
 
-This displays the following data entry fields which can be used to control the tests.
+* **Read** executes the Read action {{% todo %}}[Add a link to the read-action chapter]{{% /todo %}}
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:292]:
```suggestion
* **Read** executes the [Read action](#read)
```

File: content/appstore/connectors/opc-ua.md [hunk 5]
```diff
@@ -79,118 +93,154 @@ The module/app is designed for usage with multiple servers if necessary. The sta
...
-* Use return value – `Yes` creates a string variable containing the value read from the node, `No` does not return a variable
-* Variable name – the name assigned to the variable containing the return value
+* NodeId – The NodeId of the Node you want to read. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
 
 {{% alert type="info" %}}
-All values are read as strings, you will need to convert them if you need a numeric value.
+All values are read as strings, you will need to convert them if you need a numeric or date value.
 {{% /alert %}}
 
 #### 3.2.3 **Subscribe** to Updates of Data from a Node
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:137]:
```suggestion
#### 3.2.3 **Subscribe** to Updates of Data from a Node{#subscribe}
```

File: content/appstore/connectors/opc-ua.md [hunk 6]
```diff
@@ -219,52 +269,65 @@ This page allows you to create or change the details of an OPC UA server you wan
...
+* **Value to write** – the value to write to a node when the **Write** button is clicked. Leave blank if you want to execute any of the other actions.
+* **Result** – the resulting JSON string from performing any of the four test actions. This shows the full response from the OPC UA Server, the connector might at relevant details in case of an error. 
 
-This displays the following data entry fields which can be used to control the tests.
+* **Read** executes the Read action {{% todo %}}[Add a link to the read-action chapter]{{% /todo %}}
+* **Write** executes the Write action {{% todo %}}[Add a link to the Write-action chapter]{{% /todo %}}
+* **Browse** executes the Browse action {{% todo %}}[Add a link to the Browse-action chapter]{{% /todo %}}
+* **Subscribe Simple** executes the Subscribe action {{% todo %}}[Add a link to the Subscribe-action chapter]{{% /todo %}}. Process response with microflow: SubscriptionIncomingData_Process_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:295]:
```suggestion
* **Subscribe Simple** executes the [Subscribe action](#subscribe). Process response with microflow: SubscriptionIncomingData_Process_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
```

File: content/appstore/connectors/opc-ua.md [hunk 7]
```diff
@@ -79,118 +93,154 @@ The module/app is designed for usage with multiple servers if necessary. The sta
...
 
 * Opc ua server cfg – an object of entity type OpcUaServerCfg containing the configuration of the server to which the request is made
-* Subscription ID – the ID of the subscription from which you want to unsubscribe — this is held as the **SubscriptionID** in the **Subscription** entity
 * Monitored item ID – the ID of the item which is being monitored by the subscription — this is held as the **MonitoredItemID** in the **Subscription** entity
-* Use return value – `Yes` ????? what is the return value, `No` does not return a variable
-* Variable name – the name assigned to the variable containing the return value
+* RestartSubscriptionOnNextReboot – Indicate if the registration entities should be kept by this action. When 'true' the **MonitoredItem** & **Subscription** will be kept in the database and their Status will be changed to 'New'. When 'false' the status of the monitoredItem will become 'Deleted' and the module will automatically remove the entities from the database. 
+
 
 #### 3.2.5 **Write** Data to a Node
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:233]:
```suggestion
#### 3.2.5 **Write** Data to a Node{#write}
```

File: content/appstore/connectors/opc-ua.md [hunk 8]
```diff
@@ -219,52 +269,65 @@ This page allows you to create or change the details of an OPC UA server you wan
...
 
-To display the test buttons and additional data fields, check **Enable Test Mode**.
+**Test Messaging:**  
+* **NodeId** – The NodeId of the Node you want to use. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
+* **Value to write** – the value to write to a node when the **Write** button is clicked. Leave blank if you want to execute any of the other actions.
+* **Result** – the resulting JSON string from performing any of the four test actions. This shows the full response from the OPC UA Server, the connector might at relevant details in case of an error. 
 
-This displays the following data entry fields which can be used to control the tests.
+* **Read** executes the Read action {{% todo %}}[Add a link to the read-action chapter]{{% /todo %}}
+* **Write** executes the Write action {{% todo %}}[Add a link to the Write-action chapter]{{% /todo %}}
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:293]:
```suggestion
* **Write** executes the [Write action](#write)
```

File: content/appstore/connectors/opc-ua.md [hunk 9]
```diff
@@ -219,52 +269,65 @@ This page allows you to create or change the details of an OPC UA server you wan
...
+* **Browse** executes the Browse action {{% todo %}}[Add a link to the Browse-action chapter]{{% /todo %}}
+* **Subscribe Simple** executes the Subscribe action {{% todo %}}[Add a link to the Subscribe-action chapter]{{% /todo %}}. Process response with microflow: SubscriptionIncomingData_Process_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
+* **Subscribe w/ Msg**  executes the Subscribe action {{% todo %}}[Add a link to the Subscribe-action chapter]{{% /todo %}}. Process response with microflow: SubscriptionIncomingData_Process_wMessage_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:296]:
```suggestion
* **Subscribe w/ Msg**  executes the [Subscribe action](#subscribe). Process response with microflow: SubscriptionIncomingData_Process_wMessage_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
```

File: content/appstore/connectors/opc-ua.md [hunk 10]
```diff
@@ -38,17 +38,18 @@ The OPC UA Client connector supports all three security options offered by OPC U
 * Credentials
 * Client certificates
 
-{{% alert type="warning" %}}
-Only one of these authentication methods should be enabled for each OPC UA server. This method must be matched by the OPC UA Client connector to ensure it can reach the correct endpoint.
-{{% /alert %}}
+Only one of the options can be in use any any time, which one will be determined by the OPC UA server that you are connecting too. The password from both the user and certificate are automatically encrypted when saving the server configuration using the Encryption module.
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:41]:
```suggestion
Only one of the options can be in use any any time. Which one will be determined by the OPC UA server that you are connecting to. The password from both the user and certificate are automatically encrypted when saving the server configuration using the Encryption module.  
```

File: content/appstore/connectors/opc-ua.md [hunk 11]
```diff
@@ -38,17 +38,18 @@ The OPC UA Client connector supports all three security options offered by OPC U
...
 
 #### 1.1.3 Dependencies
 
 The OPC UA Client connector has the following dependencies
 
 * Mendix 8.8.1 or higher
+* Encryption Module
 * An OPC UA server – we assume that you are familiar with OPC UA and your OPC UA server(s), these instructions do not include information on how to set up your servers.
 
+Recommended; Having an external OPC UA Client tool will make setup of the connection a lot easier. A commonly used client is the one from [Unified Automation](https://www.unified-automation.com/downloads/opc-ua-clients.html), with this (non-Mendix) tool you will be able to browse through the server and find the parameters needed to configure your connection.
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:51]:
```suggestion
Recommended; Having an external OPC UA Client tool will make setup of the connection a lot easier. A commonly-used client is the one from [Unified Automation](https://www.unified-automation.com/downloads/opc-ua-clients.html). With this (non-Mendix) tool you will be able to browse through the server and find the parameters needed to configure your connection. 
```

File: content/appstore/connectors/opc-ua.md [hunk 12]
```diff
@@ -58,13 +59,26 @@ You will see the new module in the **App Store modules** section of the **Projec
...
 
 {{% image_container width="300" %}}
 ![OPC UA Client connector actions in the microflow toolbox](attachments/opc-ua/opc-ua-toolbox.png)
 {{% /image_container %}}
 
+
+### 2.1 Configuration:  
+1. Add the **OpcUaServer_Overview** page to the navigation of the app, either through the **Navigation** settings, or by adding an **Open Page** button to a page which is already in the navigation (for example the home page). 
+
+2. Looup the Constant: **UA_ApplicationName** in the **_USE_ME** folder and update these with the information specific to your Client. The application name must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:74]:
```suggestion
2. Find the constant: **UA_ApplicationName** in the **_USE_ME** folder and update this with the information specific to your Client. The application name must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   
```

File: content/appstore/connectors/opc-ua.md [hunk 13]
```diff
@@ -58,13 +59,26 @@ You will see the new module in the **App Store modules** section of the **Projec
...
+
+2. Looup the Constant: **UA_ApplicationName** in the **_USE_ME** folder and update these with the information specific to your Client. The application name must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   
+*See the OPC UA Specification for more details, this setting is only used by the OPC UA Server.*
+
+3. Looup the Constant: **UA_ApplicationURI** in the **_USE_ME** folder and update these with the information specific to your Client. The application URI must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).
```
Comment (MarkvanMents) [inline content/appstore/connectors/opc-ua.md:77]:
```suggestion
3. Find the constant: **UA_ApplicationURI** in the **_USE_ME** folder and update this with the information specific to your Client. The application URI must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   
```

**Review by MarkvanMents — APPROVED** [review]
Hi Jasper, @jaspervanderhoek 
I've made suggestions which add in the links which were in your todos.
When you are ready, can you accept these (if you are happy) and remove the *Draft* status of the PR. Then I can merge it.
Then I can do a final review - it will be easier as I will be able to update the files directly instead of having to make suggestions.

**Comment by MarkvanMents** [comment]
@jaspervanderhoek Jasper, I still can't merge this as you still have it in draft.
Do you have a release date for this module planned, or do I need to talk to Jos/Erno?
Thanks
Mark

**Pull request merged**